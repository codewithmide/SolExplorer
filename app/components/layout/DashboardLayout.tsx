"use client";

import { useState, useEffect } from "react";
import SideNav from "../SideNav";
import { FaGithub } from "react-icons/fa";
import ThemeSwitch from "../ThemeSwitch";
import Link from "next/link";
import { Button, Input } from "../molecules/FormComponents";
import searchIcon from "@/public/svgs/search.svg";
import LogoIcon from "@/public/svgs/logoIcon.svg";
import Image from "next/image";
import { useRouter } from "next/navigation";
import { SearchResult } from "@/app/common/types/dashboardTypes";
import DropDownSelect from "../molecules/DropDownSelect";
import { setBaseURL } from "@/app/common/utils/axios.instance";
import { useNetwork } from "@/app/common/utils/axios.instance";
import { User } from "@supabase/supabase-js";
import { supabase } from "@/app/lib/supabase";
import { FcGoogle } from "react-icons/fc";


const DashboardLayout = ({ children, path }: any) => {
  const router = useRouter();
  const { network, changeNetwork } = useNetwork();

  const [searchTerm, setSearchTerm] = useState<string>("");
  const [searchResult, setSearchResult] = useState<SearchResult | null>(null);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [activeNetwork, setActiveNetwork] = useState<string>("Mainnet");

  const [user, setUser] = useState<User | null>(null);

  useEffect(() => {
    const savedNetwork = localStorage.getItem("network") || "Mainnet";
    setActiveNetwork(savedNetwork);
    setBaseURL(savedNetwork);

    // Check for existing session
    supabase.auth.getSession().then(({ data: { session } }) => {
      setUser(session?.user ?? null);
    });

    const {
      data: { subscription },
    } = supabase.auth.onAuthStateChange((_event, session) => {
      setUser(session?.user ?? null);
    });

    return () => subscription.unsubscribe();
  }, []);

  const handleSignIn = async () => {
    try {
      const { error } = await supabase.auth.signInWithOAuth({
        provider: "google",
        options: {
          redirectTo: "https://solexplorer-v1.vercel.app/",
        },
      });
      if (error) throw error;
    } catch (error) {
      console.error("Error signing in with Google:", error);
    }
  };

  const handleSignOut = async () => {
    try {
      const { error } = await supabase.auth.signOut();
      if (error) throw error;
    } catch (error) {
      console.error("Error signing out:", error);
    }
  };

  const handleSearchChange = async (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setSearchTerm(value);
    if (value.length > 0) {
      await identifySearchType(value);
    } else {
      setSearchResult(null);
    }
  };

  // const identifySearchType = async (value: string) => {
  //   setIsLoading(true);
  //   const transactionPattern = /^[1-9A-HJ-NP-Za-km-z]{87,88}$/;
  //   const blockPattern = /^\d+$/;
  //   const accountPattern = /^[1-9A-HJ-NP-Za-km-z]{32,44}$/;

  //   try {
  //     if (blockPattern.test(value)) {
  //       setSearchResult({ type: "Block", value });
  //     } else if (transactionPattern.test(value)) {
  //       setSearchResult({ type: "Transaction", value });
  //     } else if (accountPattern.test(value)) {
  //       const accountInfo = await AccountService.fetchData("getAccountInfo", [value]);
  //       if (accountInfo && accountInfo.value !== null) {
  //         setSearchResult({ type: "Account", value });
  //       } else {
  //         const tokenInfo = await AccountService.fetchData("getTokenAccountsByOwner", [value, { programId: "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA" }]);
  //         if (tokenInfo && tokenInfo.length > 0) {
  //           setSearchResult({ type: "Token", value });
  //         } else {
  //           const programInfo = await AccountService.fetchData("getProgramAccounts", [value]);
  //           if (programInfo && programInfo.length > 0) {
  //             setSearchResult({ type: "Program", value });
  //           } else {
  //             setSearchResult({ type: "Invalid", value: "Invalid or incorrect address" });
  //           }
  //         }
  //       }
  //     } else {
  //       setSearchResult({ type: "Invalid", value: "Invalid or incorrect address" });
  //     }
  //   } catch (error) {
  //     console.error("Error identifying search type:", error);
  //     setSearchResult({ type: "Invalid", value: "Invalid or incorrect address" });
  //   } finally {
  //     setIsLoading(false);
  //   }
  // };

  const identifySearchType = (value: string) => {
    // Simple regex patterns to identify the type of search
    const accountPattern = /^[1-9A-HJ-NP-Za-km-z]{32,44}$/; // Solana account address pattern
    const transactionPattern = /^[1-9A-HJ-NP-Za-km-z]{88}$/; // Solana transaction signature pattern
    const blockPattern = /^\d+$/; // Block number pattern

    if (accountPattern.test(value)) {
      setSearchResult({ type: "Account", value });
    } else if (transactionPattern.test(value)) {
      setSearchResult({ type: "Transaction", value });
    } else if (blockPattern.test(value)) {
      setSearchResult({ type: "Block", value });
    } else {
      setSearchResult({ type: "Invalid", value: "Invalid" });
    }
  };

  const handleResultClick = () => {
    if (searchResult && searchResult.type !== "Invalid") {
      const { type, value } = searchResult;
      const path = `/dashboard/${type.toLowerCase()}/${value}`;
      router.push(path);
    }
  };

  const dropDownOptions = [
    {
      name: "Devnet",
      action: () => changeNetwork("Devnet"),
    },
    {
      name: "Mainnet",
      action: () => changeNetwork("Mainnet"),
    },
  ];

  return (
    <main className="h-[100%] overflow-hidden flex flex-col text-[#000000] relative">
      <div className="md:hidden h-full w-full absolute bg-white dark:bg-slate-800 text-black dark:text-white z-50 flex items-center justify-center">
        <div className="px-4">
          <h1 className="text-xl lg:text-2xl font-bold text-action mb-2">
            Mobile view coming soon
          </h1>
          <p className="text-base lg:text-xl max-w-xs">
            For Best experience, use a desktop device
          </p>
        </div>
      </div>

      <div className="between border-b border-[#E5E7EB] dark:border-[#374151] bg-whiteBg dark:bg-darkBg text-white-text dark:text-dark-text px-5 h-[80px]">
        <Link
          href="/"
          className="font-semibold text-teal dark:text-dark-text text-lg flex gap-1 items-center justify-center"
        >
          <Image src={LogoIcon} alt="logo" width={24} height={24} />
          SolExplorer
        </Link>

        <div className="relative">
          <Input
            preIcon={searchIcon}
            onChange={handleSearchChange}
            value={searchTerm}
            placeholder="Search transactions, account, tokens, blocks..."
            classname="w-[450px] border border-[#D1D5DB] dark:border-[#4B5563] rounded-[8px]"
          />
          {isLoading ? (
            <div className="absolute top-full left-1/2 transform -translate-x-1/2 min-w-[450px] bg-white dark:bg-slate-800 border border-[#D1D5DB] dark:border-[#4B5563] rounded-[8px] mt-1 p-4 cursor-pointer">
              <div className="animate-pulse">
                <div className="h-4 bg-gray-300 dark:bg-gray-600 rounded"></div>
                <div className="h-4 bg-gray-300 dark:bg-gray-600 rounded mt-2"></div>
              </div>
            </div>
          ) : (
            searchResult && (
              <div
                className="absolute top-full left-1/2 transform -translate-x-1/2 min-w-[450px] bg-white dark:bg-slate-800 border border-[#D1D5DB] dark:border-[#4B5563] rounded-[8px] mt-1 p-4 cursor-pointer"
                onClick={handleResultClick}
              >
                <p className="text-sm text-gray-500 dark:text-gray-400">
                  {searchResult.type}
                </p>
                <p className="text-md text-black dark:text-white">
                  {searchResult.value}
                </p>
              </div>
            )
          )}
        </div>

        <div className="flex gap-6 items-center">
          {user ? (
            <div className="flex items-center gap-2">
              <span>{user.email}</span>
              <Button
                link={handleSignOut}
                classname="px-2 bg-red-500 text-white rounded"
              >
                Sign Out
              </Button>
            </div>
          ) : (
            <Button
              link={handleSignIn}
              classname="px-3  center gap-2 border-[#E5E7EB] dark:border-[#374151] border rounded "
            >
              <p>Sign In with Google</p>
              <FcGoogle />
            </Button>
          )}
          <DropDownSelect
            cta="Select Network"
            options={dropDownOptions}
            active={network}
          />
          <ThemeSwitch />
          <a href="https://github.com/codewithmide/SolExplorer" target="_blank">
            <FaGithub size={24} />
          </a>
        </div>
      </div>

      <main
        className="w-full bg-background flex flex-row"
        style={{ height: "calc(100vh - 80px)" }}
      >
        <SideNav active={path} />
        <div className="bg-background w-full overflow-y-scroll">{children}</div>
      </main>
    </main>
  );
};

export default DashboardLayout;
