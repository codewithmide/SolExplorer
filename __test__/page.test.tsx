import React from 'react';
import { render } from '@testing-library/react';
import DashboardLayout from '@/app/components/layout/DashboardLayout';
import { useRouter } from 'next/navigation';
import '@testing-library/jest-dom';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

test('renders DashboardLayout component correctly', () => {
  (useRouter as jest.Mock).mockImplementation(() => ({
    push: jest.fn(),
    query: {},
    pathname: '/',
    asPath: '/',
  }));

  const { asFragment } = render(<DashboardLayout path="test-path">Test Children</DashboardLayout>);
  expect(asFragment()).toMatchSnapshot();
});
