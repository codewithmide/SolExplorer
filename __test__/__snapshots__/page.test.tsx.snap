// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DashboardLayout component correctly 1`] = `
<DocumentFragment>
  <main
    class="h-[100%] overflow-hidden flex flex-col text-[#000000] relative"
  >
    <div
      class="md:hidden h-full w-full absolute bg-white dark:bg-slate-800 text-black dark:text-white z-50 flex items-center justify-center"
    >
      <div
        class="px-4"
      >
        <h1
          class="text-xl lg:text-2xl font-bold text-action mb-2"
        >
          Mobile view coming soon
        </h1>
        <p
          class="text-base lg:text-xl max-w-xs"
        >
          For Best experience, use a desktop device
        </p>
      </div>
    </div>
    <div
      class="between border-b border-[#E5E7EB] dark:border-[#374151] bg-whiteBg dark:bg-darkBg text-white-text dark:text-dark-text px-5 h-[80px]"
    >
      <a
        class="font-semibold text-teal dark:text-dark-text text-lg flex gap-1 items-center justify-center"
        href="/"
      >
        <img
          alt="logo"
          data-nimg="1"
          decoding="async"
          height="24"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
          style="color: transparent;"
          width="24"
        />
        SolExplore
      </a>
      <div
        class="relative"
      >
        <div
          class="flex flex-col gap-2  w-[450px] border border-[#D1D5DB] dark:border-[#4B5563] rounded-[8px]"
        >
          <div
            class="flex items-center input-wrapper outline-none text-sm bg-[#F9FAFB] dark:bg-[#374151]  rounded-lg focus:outline-none "
          >
            <img
              alt=""
              class="h-4 w-4 ml-3"
              data-nimg="1"
              decoding="async"
              height="40"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
              style="color: transparent;"
              width="40"
            />
            <input
              class="focus:outline-none focus:bg-none p-3 placeholder-text-secondary bg-transparent w-full rounded-lg text-dark"
              placeholder="Search transactions, account, tokens, blocks..."
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="flex gap-6 items-center"
      >
        <div
          class="relative border-[#E5E7EB] dark:border-[#374151] border rounded-lg"
        >
          <button
            class="p-3 rounded-lg disabled:opacity-50 disabled:cursor-not-allowed undefined flex items-center gap-2 mb-0"
            type="submit"
          >
            <p>
              Mainnet
            </p>
          </button>
        </div>
        <svg
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 496 512"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
          />
        </svg>
      </div>
    </div>
    <main
      class="w-full bg-background flex flex-row"
      style="height: calc(100vh - 80px);"
    >
      <div
        class="w-[18rem] pt-5 bg-whiteBg dark:bg-darkBg text-white-text dark:text-dark-text"
      >
        <div
          class="sticky top-0 flex flex-col justify-between max-h-screen h-full  pl-2 pr-5 overflow-scroll"
        >
          <div>
            <div
              class="flex flex-col"
            >
              <div
                class="pl-4 py-3 cursor-pointer mb-1 text-sm flex items-start gap-2 rounded-md text-dark hover:bg-background text-[#111928] dark:text-white hover:bg-[#D5F5F6] dark:hover:bg-[#036672]"
              >
                <div
                  class="w-full "
                >
                  <div
                    class="flex items-center gap-2"
                  >
                    <img
                      alt="icon"
                      class=""
                      data-nimg="1"
                      decoding="async"
                      height="20"
                      loading="lazy"
                      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
                      srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
                      style="color: transparent;"
                      width="20"
                    />
                    <p
                      class="whitespace-nowrap"
                    >
                      Dashboard
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-col"
            >
              <div
                class="pl-4 py-3 cursor-pointer mb-1 text-sm flex items-start gap-2 rounded-md text-dark hover:bg-background text-[#111928] dark:text-white hover:bg-[#D5F5F6] dark:hover:bg-[#036672]"
              >
                <div
                  class="w-full "
                >
                  <div
                    class="flex items-center gap-2"
                  >
                    <img
                      alt="icon"
                      class=""
                      data-nimg="1"
                      decoding="async"
                      height="20"
                      loading="lazy"
                      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
                      srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
                      style="color: transparent;"
                      width="20"
                    />
                    <p
                      class="whitespace-nowrap"
                    >
                      Accounts
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-col"
            >
              <div
                class="pl-4 py-3 cursor-pointer mb-1 text-sm flex items-start gap-2 rounded-md text-dark hover:bg-background text-[#111928] dark:text-white hover:bg-[#D5F5F6] dark:hover:bg-[#036672]"
              >
                <div
                  class="w-full "
                >
                  <div
                    class="flex items-center gap-2"
                  >
                    <img
                      alt="icon"
                      class=""
                      data-nimg="1"
                      decoding="async"
                      height="20"
                      loading="lazy"
                      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
                      srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
                      style="color: transparent;"
                      width="20"
                    />
                    <p
                      class="whitespace-nowrap"
                    >
                      Transactions
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-col"
            >
              <div
                class="pl-4 py-3 cursor-pointer mb-1 text-sm flex items-start gap-2 rounded-md text-dark hover:bg-background text-[#111928] dark:text-white hover:bg-[#D5F5F6] dark:hover:bg-[#036672]"
              >
                <div
                  class="w-full "
                >
                  <div
                    class="flex items-center gap-2"
                  >
                    <img
                      alt="icon"
                      class=""
                      data-nimg="1"
                      decoding="async"
                      height="20"
                      loading="lazy"
                      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
                      srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
                      style="color: transparent;"
                      width="20"
                    />
                    <p
                      class="whitespace-nowrap"
                    >
                      Blocks
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-col"
            >
              <div
                class="pl-4 py-3 cursor-pointer mb-1 text-sm flex items-start gap-2 rounded-md text-dark hover:bg-background text-[#111928] dark:text-white hover:bg-[#D5F5F6] dark:hover:bg-[#036672]"
              >
                <div
                  class="w-full "
                >
                  <div
                    class="flex items-center gap-2"
                  >
                    <img
                      alt="icon"
                      class=""
                      data-nimg="1"
                      decoding="async"
                      height="20"
                      loading="lazy"
                      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
                      srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
                      style="color: transparent;"
                      width="20"
                    />
                    <p
                      class="whitespace-nowrap"
                    >
                      Tokens
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-col"
            >
              <div
                class="pl-4 py-3 cursor-pointer mb-1 text-sm flex items-start gap-2 rounded-md text-dark hover:bg-background text-[#111928] dark:text-white hover:bg-[#D5F5F6] dark:hover:bg-[#036672]"
              >
                <div
                  class="w-full "
                >
                  <div
                    class="flex items-center gap-2"
                  >
                    <img
                      alt="icon"
                      class=""
                      data-nimg="1"
                      decoding="async"
                      height="20"
                      loading="lazy"
                      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
                      srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
                      style="color: transparent;"
                      width="20"
                    />
                    <p
                      class="whitespace-nowrap"
                    >
                      NFTs
                    </p>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-col"
            >
              <div
                class="pl-4 py-3 cursor-pointer mb-1 text-sm flex items-start gap-2 rounded-md text-dark hover:bg-background text-[#111928] dark:text-white hover:bg-[#D5F5F6] dark:hover:bg-[#036672]"
              >
                <div
                  class="w-full "
                >
                  <div
                    class="flex items-center gap-2"
                  >
                    <img
                      alt="icon"
                      class=""
                      data-nimg="1"
                      decoding="async"
                      height="20"
                      loading="lazy"
                      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
                      srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
                      style="color: transparent;"
                      width="20"
                    />
                    <p
                      class="whitespace-nowrap"
                    >
                      Developer Tools
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="flex flex-col pb-6 pl-2"
          >
            <div
              class="flex gap-2"
            >
              <img
                alt="Solana"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
              <p
                class="font-semibold text-sm"
              >
                Powered by Solana
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="bg-background w-full overflow-y-scroll"
      >
        Test Children
      </div>
    </main>
  </main>
</DocumentFragment>
`;
